source/LinkedList.inl

```diff
@@ -81,9 +69,54 @@ DataStructures::LinkedList<T>::LinkedList(std::initializer_list<T> contents)
 template<typename T>
 DataStructures::LinkedList<T>::LinkedList()
 {
-  assert_type_has_operator_equal_and_nequal<T>();
   number_of_elements = 0;
   first = nullptr;
+  last = nullptr;
+}
+
+template<typename T>
+DataStructures::LinkedList<T>::LinkedList(const LinkedList& other)
+{
+  *this = other;
+}
+
+template<typename T>
+DataStructures::LinkedList<T>::LinkedList(LinkedList&& other) noexcept
+{
+  *this = std::move(other);
+}
+
+template<typename T>
+LinkedList<T>&
+DataStructures::LinkedList<T>::operator=(const LinkedList& other)
+{
+  first = nullptr;
+  number_of_elements = other.size();
+  Element* previous;
+  for (auto it = other.begin(); it != other.end(); ++it) {
+    Element* next = new Element{ *it, nullptr };
+    if (first == nullptr) {
+      first = next;
+    } else {
+      previous->next = next;
+    }
+    previous = next;
+  }
+  last = previous;
+  return *this;
+}
+
+template<typename T>
+LinkedList<T>&
+DataStructures::LinkedList<T>::operator=(LinkedList&& other) noexcept
+{
+  first = std::move(other.first);
```

Comment:
This is very close! The one case you miss is when the target list has data that need to be destructed. If you have lists `a` and `b`, add data to both and then move `b` in to `a`, you'll leak the data previously in `a`;

A pattern that I've used that I think is good is defining move operations in terms of `std::swap`. Give your class a swap friend function (such that std::swap can find it), make sure it swaps *everything*, and then just call that in your move ctor/assign operator.

Reply:
why would `other` have data to destruct?  The `first` and `last` pointers are being moved, which nullifies destroys them in `other`, and the rest of the elements will rightly be managed by `this`.  Otherwise, there's just `number_of_elements` which doesn't need to be explicitly destructed

Reply:
Here's an example:

```
class BadMove {
private:
    int* ptr = nullptr;
public:
    virtual ~BadMove() {
        if (nullptr != ptr) {
            delete ptr;
            ptr = nullptr;
        }
    }
    BadMove() = default;
    BadMove(int x) : ptr{new int(x)} {}
    BadMove(const BadMove&) = delete;
    BadMove(BadMove&& other) {
        ptr = other.ptr;
        other.ptr = nullptr;
    }

    BadMove& operator=(const BadMove&) = delete;
    BadMove& operator=(BadMove&& other) {
        ptr = other.ptr;
        other.ptr =  nullptr;
        return *this;
    }
};

int main() {
    BadMove a{5};
    BadMove b{6};
    a = std::move(b);

    return 0;
}
```

`a` and `b` both have data before the move. After the move, `a` has `b`'s old pointer which it will know to delete, but `a`'s old pointer is leaked.